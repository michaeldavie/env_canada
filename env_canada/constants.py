USER_AGENT = "env_canada/0.6.2"
