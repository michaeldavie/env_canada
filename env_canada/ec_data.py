import re
import xml.etree.ElementTree as et

from geopy import distance
from ratelimit import limits, RateLimitException
import requests

SITE_LIST_URL = 'https://dd.weather.gc.ca/citypage_weather/docs/site_list_en.csv'
AQHI_SITE_LIST_URL = 'https://dd.weather.gc.ca/air_quality/doc/AQHI_XML_File_List.xml'

WEATHER_URL = 'https://dd.weather.gc.ca/citypage_weather/xml/{}_{}.xml'
AQHI_OBSERVATION_URL = 'https://dd.weather.gc.ca/air_quality/aqhi/{}/observation/realtime/xml/AQ_OBS_{}_CURRENT.xml'
AQHI_FORECAST_URL = 'https://dd.weather.gc.ca/air_quality/aqhi/{}/forecast/realtime/xml/AQ_FCST_{}_CURRENT.xml'

conditions_meta = {
    'temperature': {
        'xpath': './currentConditions/temperature',
        'english': 'Temperature',
        'french': 'Température'
    },
    'dewpoint': {
        'xpath': './currentConditions/dewpoint',
        'english': 'Dew Point',
        'french': 'Point de rosée'
    },
    'wind_chill': {
        'xpath': './currentConditions/windChill',
        'english': 'Wind Chill',
        'french': 'Refroidissement éolien'
    },
    'humidex': {
        'xpath': './currentConditions/humidex',
        'english': 'Humidex',
        'french': 'Humidex'
    },
    'pressure': {
        'xpath': './currentConditions/pressure',
        'english': 'Pressure',
        'french': 'Pression'
    },
    'tendency': {
        'xpath': './currentConditions/pressure',
        'attribute': 'tendency',
        'english': 'Tendency',
        'french': 'Tendance'
    },
    'humidity': {
        'xpath': './currentConditions/relativeHumidity',
        'english': 'Humidity',
        'french': 'Humidité'
    },
    'visibility': {
        'xpath': './currentConditions/visibility',
        'english': 'Visibility',
        'french': 'Visibilité'
    },
    'condition': {
        'xpath': './currentConditions/condition',
        'english': 'Condition',
        'french': 'Condition'
    },
    'wind_speed': {
        'xpath': './currentConditions/wind/speed',
        'english': 'Wind Speed',
        'french': 'Vitesse de vent'
    },
    'wind_gust': {
        'xpath': './currentConditions/wind/gust',
        'english': 'Wind Gust',
        'french': 'Rafale de vent'
    },
    'wind_dir': {
        'xpath': './currentConditions/wind/direction',
        'english': 'Wind Direction',
        'french': 'Direction de vent'
    },
    'wind_bearing': {
        'xpath': './currentConditions/wind/bearing',
        'english': 'Wind Bearing',
        'french': 'Palier de vent'
    },
    'high_temp': {
        'xpath': './forecastGroup/forecast/temperatures/temperature[@class="high"]',
        'english': 'High Temperature',
        'french': 'Haute température'
    },
    'low_temp': {
        'xpath': './forecastGroup/forecast/temperatures/temperature[@class="low"]',
        'english': 'Low Temperature',
        'french': 'Basse température'
    },
    'uv_index': {
        'xpath': './forecastGroup/forecast/uv/index',
        'english': 'UV Index',
        'french': 'Indice UV'
    },
    'pop': {
        'xpath': './forecastGroup/forecast/abbreviatedForecast/pop',
        'english': 'Chance of Precip.',
        'french': 'Probabilité d\'averses'
    },
    'icon_code': {
        'xpath': './currentConditions/iconCode',
        'english': 'Icon Code',
        'french': 'Code icône'
    },
    'precip_yesterday': {
        'xpath': './yesterdayConditions/precip',
        'english': 'Precipitation Yesterday',
        'french': 'Précipitation d\'hier'
    },
}

aqhi_meta = {
    'label': {
        'english': 'Air Quality Health Index',
        'french': 'Cote air santé'
    }
}

summary_meta = {
    'forecast_period': {
        'xpath': './forecastGroup/forecast/period',
        'attribute': 'textForecastName',
    },
    'text_summary': {
        'xpath': './forecastGroup/forecast/textSummary',
    },
    'label': {
        'english': 'Forecast',
        'french': 'Prévision'
    }
}

alerts_meta = {
    'warnings': {
        'english': {
            'label': 'Warnings',
            'pattern': '.*WARNING((?!ENDED).)*$'
        },
        'french': {
            'label': 'Alertes',
            'pattern': '.*(ALERTE|AVERTISSEMENT)((?!TERMINÉ).)*$'
        }
    },
    'watches': {
        'english': {
            'label': 'Watches',
            'pattern': '.*WATCH((?!ENDED).)*$'
        },
        'french': {
            'label': 'Veilles',
            'pattern': '.*VEILLE((?!TERMINÉ).)*$'
        }
    },
    'advisories': {
        'english': {
            'label': 'Advisories',
            'pattern': '.*ADVISORY((?!ENDED).)*$'
        },
        'french': {
            'label': 'Avis',
            'pattern': '.*AVIS((?!TERMINÉ).)*$'
        }
    },
    'statements': {
        'english': {
            'label': 'Statements',
            'pattern': '.*STATEMENT((?!ENDED).)*$'
        },
        'french': {
            'label': 'Bulletins',
            'pattern': '.*BULLETIN((?!TERMINÉ).)*$'
        }
    },
    'endings': {
        'english': {
            'label': 'Endings',
            'pattern': '.*ENDED'
        },
        'french': {
            'label': 'Terminaisons',
            'pattern': '.*TERMINÉE?'
        }
    }
}

metadata_meta = {
    'timestamp': {
        'xpath': './currentConditions/dateTime/timeStamp',
    },
    'location': {
        'xpath': './location/name',
    },
    'station': {
        'xpath': './currentConditions/station',
    },
}


def ignore_ratelimit_error(fun):
    def res(*args, **kwargs):
        try:
            return fun(*args, **kwargs)
        except RateLimitException:
            return None
    return res


class ECData(object):

    """Get weather data from Environment Canada."""

    def __init__(self,
                 station_id=None,
                 coordinates=None,
                 language='english'):
        """Initialize the data object."""
        self.language = language
        self.language_abr = language[:2].upper()
        self.zone_name_tag = 'name_%s_CA' % self.language_abr.lower()
        self.region_name_tag = 'name%s' % self.language_abr.title()

        self.metadata = {}
        self.conditions = {}
        self.alerts = {}
        self.daily_forecasts = []
        self.hourly_forecasts = []
        self.aqhi = {}
        self.forecast_time = ''
        self.aqhi_id = None

        if station_id:
            self.station_id = station_id
        else:
            self.station_id = self.closest_site(coordinates[0],
                                                coordinates[1])

        self.update()

    @ignore_ratelimit_error
    @limits(calls=2, period=60)
    def update(self):
        """Get the latest data from Environment Canada."""
        weather_result = requests.get(WEATHER_URL.format(self.station_id,
                                                         self.language[0]),
                                      timeout=10)
        weather_xml = weather_result.content.decode('iso-8859-1')
        weather_tree = et.fromstring(weather_xml)

        # Update metadata
        for m, meta in metadata_meta.items():
            element = weather_tree.find(meta['xpath'])
            if element is not None:
                self.metadata[m] = weather_tree.find(meta['xpath']).text
            else:
                self.metadata[m] = None

        # Update current conditions
        def get_condition(meta):
            condition = {}

            element = weather_tree.find(meta['xpath'])

            if element is not None:
                if meta.get('attribute'):
                    condition['value'] = element.attrib.get(meta['attribute'])
                else:
                    condition['value'] = element.text
                    if element.attrib.get('units'):
                        condition['unit'] = element.attrib.get('units')
            return condition

        for c, meta in conditions_meta.items():
            self.conditions[c] = {'label': meta[self.language]}
            self.conditions[c].update(get_condition(meta))

        # Update text summary
        period = get_condition(summary_meta['forecast_period'])['value']
        summary = get_condition(summary_meta['text_summary'])['value']

        self.conditions['text_summary'] = {
            'label': summary_meta['label'][self.language],
            'value': '. '.join([period, summary])
        }

        # Update alerts
        for category, meta in alerts_meta.items():
            self.alerts[category] = {'value': [],
                                     'label': meta[self.language]['label']}

        alert_elements = weather_tree.findall('./warnings/event')

        for a in alert_elements:
            title = a.attrib.get('description').strip()
            for category, meta in alerts_meta.items():
                category_match = re.search(meta[self.language]['pattern'], title)
                if category_match:
                    alert = {'title': title.title(),
                             'date': a.find('./dateTime[last()]/textSummary').text,
                             }
                    self.alerts[category]['value'].append(alert)

        # Update daily forecasts
        self.forecast_time = weather_tree.findtext('./forecastGroup/dateTime/timeStamp')
        self.daily_forecasts = []
        self.hourly_forecasts = []

        for f in weather_tree.findall('./forecastGroup/forecast'):
            self.daily_forecasts.append({
                'period': f.findtext('period'),
                'text_summary': f.findtext('textSummary'),
                'icon_code': f.findtext('./abbreviatedForecast/iconCode'),
                'temperature': f.findtext('./temperatures/temperature'),
                'temperature_class': f.find('./temperatures/temperature').attrib.get('class'),
                'precip_probability': f.findtext('./abbreviatedForecast/pop') or "0"
            })

        # Update hourly forecasts
        for f in weather_tree.findall('./hourlyForecastGroup/hourlyForecast'):
            self.hourly_forecasts.append({
                'period': f.attrib.get('dateTimeUTC'),
                'condition': f.findtext('./condition'),
                'temperature': f.findtext('./temperature'),
                'icon_code': f.findtext('./iconCode'),
                'precip_probability': f.findtext('./lop')  or "0",
            })

        # Update AQHI current condition

        if self.aqhi_id is None:
            lat = weather_tree.find('./location/name').attrib.get('lat')[:-1]
            lon = weather_tree.find('./location/name').attrib.get('lon')[:-1]
            aqhi_coordinates = (float(lat), float(lon) * -1)
            self.aqhi_id = self.closest_aqhi(aqhi_coordinates[0], aqhi_coordinates[1])

        aqhi_result = requests.get(AQHI_OBSERVATION_URL.format(self.aqhi_id[0],
                                                               self.aqhi_id[1]),
                                   timeout=10)
        if aqhi_result.status_code == 404:
            self.aqhi['current'] = None
        else:
            aqhi_xml = aqhi_result.content.decode("utf-8")
            aqhi_tree = et.fromstring(aqhi_xml)

            element = aqhi_tree.find('airQualityHealthIndex')
            if element is not None:
                self.aqhi['current'] = element.text
            else:
                self.aqhi['current'] = None

            self.conditions['air_quality'] = {
                'label': aqhi_meta['label'][self.language],
                'value': self.aqhi['current']
            }

            element = aqhi_tree.find('./dateStamp/UTCStamp')
            if element is not None:
                self.aqhi['utc_time'] = element.text
            else:
                self.aqhi['utc_time'] = None

        # Update AQHI forecasts
        aqhi_result = requests.get(AQHI_FORECAST_URL.format(self.aqhi_id[0],
                                                            self.aqhi_id[1]),
                                   timeout=10)
        if aqhi_result.status_code == 404:
            self.aqhi['forecasts'] = None
        else:
            aqhi_xml = aqhi_result.content.decode("ISO-8859-1")
            aqhi_tree = et.fromstring(aqhi_xml)

            self.aqhi['forecasts'] = {'daily': [],
                                      'hourly': []}

            # Update AQHI daily forecasts
            period = None
            for f in aqhi_tree.findall("./forecastGroup/forecast"):
                for p in f.findall("./period"):
                    if self.language_abr == p.attrib["lang"]:
                        period = p.attrib["forecastName"]
                self.aqhi['forecasts']['daily'].append(
                    {
                        "period": period,
                        "aqhi": f.findtext("./airQualityHealthIndex"),
                    }
                )

            # Update AQHI hourly forecasts
            for f in aqhi_tree.findall("./hourlyForecastGroup/hourlyForecast"):
                self.aqhi['forecasts']['hourly'].append(
                    {"period": f.attrib["UTCTime"], "aqhi": f.text}
                )

    def get_ec_sites(self):
        """Get list of all sites from Environment Canada, for auto-config."""
        import csv
        import io

        sites = []

        sites_csv_string = requests.get(SITE_LIST_URL, timeout=10).text
        sites_csv_stream = io.StringIO(sites_csv_string)

        sites_csv_stream.seek(0)
        next(sites_csv_stream)

        sites_reader = csv.DictReader(sites_csv_stream)

        for site in sites_reader:
            if site['Province Codes'] != 'HEF':
                site['Latitude'] = float(site['Latitude'].replace('N', ''))
                site['Longitude'] = -1 * float(site['Longitude'].replace('W', ''))
                sites.append(site)

        return sites

    def closest_site(self, lat, lon):
        """Return the province/site_code of the closest station to our lat/lon."""
        site_list = self.get_ec_sites()

        def site_distance(site):
            """Calculate distance to a site."""
            return distance.distance((lat, lon), (site['Latitude'], site['Longitude']))

        closest = min(site_list, key=site_distance)

        return '{}/{}'.format(closest['Province Codes'], closest['Codes'])

    def get_aqhi_regions(self):
        """Get list of all AQHI regions from Environment Canada, for auto-config."""
        result = requests.get(AQHI_SITE_LIST_URL, timeout=10)
        site_xml = result.content.decode("utf-8")
        xml_object = et.fromstring(site_xml)

        regions = []
        for zone in xml_object.findall("./EC_administrativeZone"):
            _zone_attribs = zone.attrib
            _zone_attrib = {
                "abbreviation": _zone_attribs["abreviation"],
                "zone_name": _zone_attribs[self.zone_name_tag],
            }
            for region in zone.findall("./regionList/region"):
                _region_attribs = region.attrib

                _region_attrib = {"region_name": _region_attribs[self.region_name_tag],
                                  "cgndb": _region_attribs["cgndb"],
                                  "latitude": float(_region_attribs["latitude"]),
                                  "longitude": float(_region_attribs["longitude"])}
                _children = region
                for child in _children:
                    _region_attrib[child.tag] = child.text
                _region_attrib.update(_zone_attrib)
                regions.append(_region_attrib)
        return regions

    def closest_aqhi(self, lat, lon):
        """Return the AQHI region and site ID of the closest site."""
        region_list = self.get_aqhi_regions()

        def site_distance(site):
            """Calculate distance to a region."""
            return distance.distance(
                (lat, lon), (site["latitude"], site["longitude"])
            )
        closest = min(region_list, key=site_distance)

        return closest['abbreviation'], closest['cgndb']
