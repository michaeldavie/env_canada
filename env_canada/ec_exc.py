class UnknownStationId(Exception):
    """Station ID is unknown"""
