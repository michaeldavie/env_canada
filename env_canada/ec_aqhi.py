from datetime import datetime, timezone
import logging

from aiohttp import ClientSession
import defusedxml.ElementTree as et
from geopy import distance
import voluptuous as vol

from .constants import USER_AGENT

AQHI_SITE_LIST_URL = "https://dd.weather.gc.ca/air_quality/doc/AQHI_XML_File_List.xml"
AQHI_OBSERVATION_URL = "https://dd.weather.gc.ca/air_quality/aqhi/{}/observation/realtime/xml/AQ_OBS_{}_CURRENT.xml"
AQHI_FORECAST_URL = "https://dd.weather.gc.ca/air_quality/aqhi/{}/forecast/realtime/xml/AQ_FCST_{}_CURRENT.xml"

LOG = logging.getLogger(__name__)

ATTRIBUTION = {
    "EN": "Data provided by Environment Canada",
    "FR": "Donn√©es fournies par Environnement Canada",
}


__all__ = ["ECAirQuality"]


def timestamp_to_datetime(timestamp):
    dt = datetime.strptime(timestamp, "%Y%m%d%H%M%S")
    dt = dt.replace(tzinfo=timezone.utc)
    return dt


async def get_aqhi_regions(language):
    """Get list of all AQHI regions from Environment Canada, for auto-config."""
    zone_name_tag = "name_%s_CA" % language.lower()
    region_name_tag = "name%s" % language.title()

    LOG.debug("get_aqhi_regions() started")

    regions = []
    async with ClientSession(raise_for_status=True) as session:
        response = await session.get(
            AQHI_SITE_LIST_URL, headers={"User-Agent": USER_AGENT}, timeout=10
        )
        result = await response.read()

    site_xml = result.decode("utf-8")
    xml_object = et.fromstring(site_xml)

    for zone in xml_object.findall("./EC_administrativeZone"):
        _zone_attribs = zone.attrib
        _zone_attrib = {
            "abbreviation": _zone_attribs["abreviation"],
            "zone_name": _zone_attribs[zone_name_tag],
        }
        for region in zone.findall("./regionList/region"):
            _region_attribs = region.attrib

            _region_attrib = {
                "region_name": _region_attribs[region_name_tag],
                "cgndb": _region_attribs["cgndb"],
                "latitude": float(_region_attribs["latitude"]),
                "longitude": float(_region_attribs["longitude"]),
            }
            _children = list(region)
            for child in _children:
                _region_attrib[child.tag] = child.text
            _region_attrib.update(_zone_attrib)
            regions.append(_region_attrib)

    LOG.debug("get_aqhi_regions(): found %d regions", len(regions))

    return regions


async def find_closest_region(language, lat, lon):
    """Return the AQHI region and site ID of the closest site."""
    region_list = await get_aqhi_regions(language)

    def site_distance(site):
        """Calculate distance to a region."""
        return distance.distance((lat, lon), (site["latitude"], site["longitude"]))

    return min(region_list, key=site_distance)


class ECAirQuality(object):

    """Get air quality data from Environment Canada."""

    def __init__(self, **kwargs):
        """Initialize the data object."""

        init_schema = vol.Schema(
            vol.All(
                vol.Any(
                    {
                        vol.Required("coordinates"): object,
                        vol.Optional("language"): object,
                    },
                    {
                        vol.Required("zone_id"): object,
                        vol.Required("region_id"): object,
                        vol.Optional("language"): object,
                    },
                ),
                {
                    vol.Optional("zone_id"): vol.In(
                        ["atl", "ont", "pnr", "pyr", "que"]
                    ),
                    vol.Optional("region_id"): vol.All(str, vol.Length(5)),
                    vol.Optional("coordinates"): (
                        vol.All(vol.Or(int, float), vol.Range(-90, 90)),
                        vol.All(vol.Or(int, float), vol.Range(-180, 180)),
                    ),
                    vol.Optional("language", default="EN"): vol.In(["EN", "FR"]),
                },
            )
        )

        kwargs = init_schema(kwargs)

        self.language = kwargs["language"]

        if (
            "zone_id" in kwargs
            and "region_id" in kwargs
            and kwargs["zone_id"] is not None
            and kwargs["region_id"] is not None
        ):
            self.zone_id = kwargs["zone_id"]
            self.region_id = kwargs["region_id"].upper()
        else:
            self.zone_id = None
            self.region_id = None
            self.coordinates = kwargs["coordinates"]

        self.metadata = {"attribution": ATTRIBUTION[self.language]}
        self.region_name = None
        self.current = None
        self.current_timestamp = None
        self.forecasts = dict(daily={}, hourly={})

    async def get_aqhi_data(self, url):
        async with ClientSession(raise_for_status=True) as session:
            try:
                response = await session.get(
                    url.format(self.zone_id, self.region_id),
                    headers={"User-Agent": USER_AGENT},
                    timeout=10,
                )
            except Exception:
                LOG.debug("Retrieving AQHI failed", exc_info=True)
                return None

            result = await response.read()
            aqhi_xml = result.decode("ISO-8859-1")
            return et.fromstring(aqhi_xml)

    async def update(self):
        # Find closest site if not identified

        if not (self.zone_id and self.region_id):
            closest = await find_closest_region(self.language, *self.coordinates)
            self.zone_id = closest["abbreviation"]
            self.region_id = closest["cgndb"]
            LOG.debug(
                "update() closest region returned: zone_id '%s' region_id '%s'",
                self.zone_id,
                self.region_id,
            )

        # Fetch current measurement
        aqhi_current = await self.get_aqhi_data(url=AQHI_OBSERVATION_URL)

        if aqhi_current:
            # Update region name
            element = aqhi_current.find("region")
            self.region_name = element.attrib[
                "name{lang}".format(lang=self.language.title())
            ]
            self.metadata["location"] = self.region_name

            # Update AQHI current condition
            element = aqhi_current.find("airQualityHealthIndex")
            if element is not None:
                self.current = float(element.text)
            else:
                self.current = None

            element = aqhi_current.find("./dateStamp/UTCStamp")
            if element is not None:
                self.current_timestamp = timestamp_to_datetime(element.text)
            else:
                self.current_timestamp = None
            self.metadata["timestamp"] = self.current_timestamp
            LOG.debug(
                "update(): aqhi_current %d timestamp %s",
                self.current,
                self.current_timestamp,
            )

        # Update AQHI forecasts
        aqhi_forecast = await self.get_aqhi_data(url=AQHI_FORECAST_URL)

        if aqhi_forecast:
            # Update AQHI daily forecasts
            for f in aqhi_forecast.findall("./forecastGroup/forecast"):
                for p in f.findall("./period"):
                    if self.language == p.attrib["lang"]:
                        period = p.attrib["forecastName"]
                self.forecasts["daily"][period] = int(
                    f.findtext("./airQualityHealthIndex") or 0
                )

            # Update AQHI hourly forecasts
            for f in aqhi_forecast.findall("./hourlyForecastGroup/hourlyForecast"):
                self.forecasts["hourly"][
                    timestamp_to_datetime(f.attrib["UTCTime"])
                ] = int(f.text or 0)
