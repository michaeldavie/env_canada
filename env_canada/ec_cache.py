from aiohttp import ClientSession
from datetime import datetime, timedelta

from .constants import USER_AGENT

CACHE_EXPIRE_TIME = timedelta(minutes=200)  # Time is tuned for 3h radar image


class CacheClientSession(ClientSession):
    """Shim to cache ClientSession requests."""

    _cache = {}

    def _flush_cache(self):
        """Flush expired cache entries."""

        now = datetime.now()
        expired = [key for key, value in self._cache.items() if value[0] < now]
        for key in expired:
            del self._cache[key]

    async def get(self, url, params, cache_time=CACHE_EXPIRE_TIME):
        """Thin wrapper around ClientSession.get to cache responses."""

        self._flush_cache()  # Flush at start so we don't use expired entries

        cache_key = (url, tuple(sorted(params.items())))
        result = self._cache.get(cache_key)
        if not result:
            result = (
                datetime.now() + cache_time,
                await super().get(
                    url=url, params=params, headers={"User-Agent": USER_AGENT}
                ),
            )
            self._cache[cache_key] = result

        return result[1]
