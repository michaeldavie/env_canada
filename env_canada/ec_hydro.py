import csv
import io

from aiohttp import ClientSession
from dateutil.parser import isoparse
from geopy import distance
import voluptuous as vol

from .constants import USER_AGENT

SITE_LIST_URL = "https://dd.weather.gc.ca/hydrometric/doc/hydrometric_StationList.csv"
READINGS_URL = "https://dd.weather.gc.ca/hydrometric/csv/{prov}/hourly/{prov}_{station}_hourly_hydrometric.csv"


__all__ = ["ECHydro"]


async def get_hydro_sites():
    """Get list of all sites from Environment Canada, for auto-config."""

    sites = []

    async with ClientSession(raise_for_status=True) as session:
        response = await session.get(
            SITE_LIST_URL, headers={"User-Agent": USER_AGENT}, timeout=10
        )
        result = await response.read()
    sites_csv_string = result.decode("utf-8-sig")
    sites_csv_stream = io.StringIO(sites_csv_string)

    header = [h.split("/")[0].strip() for h in sites_csv_stream.readline().split(",")]
    sites_reader = csv.DictReader(sites_csv_stream, fieldnames=header)

    for site in sites_reader:
        # Ignore bad site data
        if site["Latitude"] is not None:
            site["Latitude"] = float(site["Latitude"])
            site["Longitude"] = float(site["Longitude"])
            sites.append(site)

    return sites


async def closest_site(lat, lon):
    """Return the province/site_code of the closest station to our lat/lon."""
    site_list = await get_hydro_sites()

    def site_distance(site):
        """Calculate distance to a site."""
        return distance.distance((lat, lon), (site["Latitude"], site["Longitude"]))

    closest = min(site_list, key=site_distance)

    return closest


class ECHydro(object):

    """Get hydrometric data from Environment Canada."""

    def __init__(self, **kwargs):
        """Initialize the data object."""

        init_schema = vol.Schema(
            vol.All(
                vol.Any(
                    {
                        vol.Required("coordinates"): object,
                    },
                    {
                        vol.Required("province"): object,
                        vol.Required("station"): object,
                    },
                ),
                {
                    vol.Optional("province"): vol.All(str, vol.Length(2)),
                    vol.Optional("station"): vol.All(str, vol.Length(7)),
                    vol.Optional("coordinates"): (
                        vol.All(vol.Or(int, float), vol.Range(-90, 90)),
                        vol.All(vol.Or(int, float), vol.Range(-180, 180)),
                    ),
                },
            )
        )

        kwargs = init_schema(kwargs)

        self.measurements = {}
        self.timestamp = None
        self.location = None

        if (
            "province" in kwargs
            and "station" in kwargs
            and kwargs["province"] is not None
            and kwargs["station"] is not None
        ):
            self.province = kwargs["province"]
            self.station = kwargs["station"]
        else:
            self.province = None
            self.station = None
            self.coordinates = kwargs["coordinates"]

    async def update(self):
        """Get the latest data from Environment Canada."""

        # Determine closest site if not provided

        if not (self.province and self.station) and self.coordinates:
            closest = await closest_site(*self.coordinates)
            self.province = closest["Prov"]
            self.station = closest["ID"]
            self.location = closest["Name"].title()

        # Get hydrometric data

        async with ClientSession(raise_for_status=True) as session:
            response = await session.get(
                READINGS_URL.format(prov=self.province, station=self.station),
                headers={"User-Agent": USER_AGENT},
                timeout=10,
            )
            result = await response.read()
        hydro_csv_string = result.decode("utf-8-sig")
        hydro_csv_stream = io.StringIO(hydro_csv_string)

        header = [
            h.split("/")[0].strip() for h in hydro_csv_stream.readline().split(",")
        ]
        readings_reader = csv.DictReader(hydro_csv_stream, fieldnames=header)

        readings = [r for r in readings_reader]
        if len(readings) > 0:
            latest = readings[-1]

            if latest["Water Level"] != "":
                self.measurements["water_level"] = {
                    "label": "Water Level",
                    "value": float(latest["Water Level"]),
                    "unit": "m",
                }

            if latest["Discharge"] != "":
                self.measurements["discharge"] = {
                    "label": "Discharge",
                    "value": float(latest["Discharge"]),
                    "unit": "mÂ³/s",
                }

            self.timestamp = isoparse(readings[-1]["Date"])
