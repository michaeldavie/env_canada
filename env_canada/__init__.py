from .ec_aqhi import ECAirQuality
from .ec_historical import ECHistorical, ECHistoricalRange
from .ec_hydro import ECHydro
from .ec_radar import ECRadar
from .ec_weather import ECWeather
