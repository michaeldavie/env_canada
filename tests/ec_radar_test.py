import asyncio
from datetime import date, datetime
from io import BytesIO
from unittest.mock import AsyncMock, patch

import pytest
from PIL import Image
from syrupy.assertion import SnapshotAssertion

from env_canada import ECRadar


@pytest.mark.slow
@pytest.mark.parametrize(
    "init_parameters",
    [
        {"coordinates": (50, -100), "precip_type": "snow", "legend": False},
        {"coordinates": (50, -100), "precip_type": "rain", "timestamp": False},
        {"coordinates": (50, -100)},
        {"coordinates": (50, -100), "precip_type": None},
    ],
)
def test_ecradar(init_parameters):
    radar = ECRadar(**init_parameters)
    frame = asyncio.run(radar.get_latest_frame())
    image = Image.open(BytesIO(frame))
    assert image.format == "PNG"


@pytest.fixture
def test_radar():
    return ECRadar(coordinates=(50, -100))


@pytest.mark.slow
def test_get_dimensions(test_radar):
    dimensions = asyncio.run(test_radar._get_dimensions())
    assert isinstance(dimensions[0], datetime) and isinstance(dimensions[1], datetime)


@pytest.mark.slow
def test_get_latest_frame(test_radar):
    frame = asyncio.run(test_radar.get_latest_frame())
    image = Image.open(BytesIO(frame))
    assert image.format == "PNG"


@pytest.mark.slow
def test_get_loop(test_radar):
    loop = asyncio.run(test_radar.get_loop())
    image = Image.open(BytesIO(loop))
    assert image.format == "GIF" and image.is_animated


def test_set_precip_type(test_radar):
    test_radar.precip_type = "auto"
    assert test_radar.precip_type[0] == "auto"

    if date.today().month in range(4, 11):
        assert test_radar.precip_type[1] == "rain"
    else:
        assert test_radar.precip_type[1] == "snow"


@pytest.mark.asyncio
async def test_get_radar_image_with_mock_data(snapshot: SnapshotAssertion):
    """
    This is still technically a slow test, in that it uses a lot of CPU.
    Still useful as it is a complete test of the entire radar code
    except for going to the network.
    """

    def mock_get_resource(_, params, bytes=True):
        fname = f"tests/fixtures/radar/{params['request']}_{params.get('time', '')}"
        with open(fname, "rb") as f:
            return f.read()

    # Using 800x800 this test works locally but fails in GitHub CI. Using 100x100,
    # the test works in both places. Consider this fragile, but we at least have
    # a local test. It's unclear why the test doesn't work at the bigger size
    # in CI. Best guess is PIL library generates a different image based on CPU
    # architecture. If need be the image can be ignored in the snapshot compare.
    tr = ECRadar(coordinates=(50, -100), width=100, height=100)
    with patch(
        "env_canada.ec_radar._get_resource", AsyncMock(side_effect=mock_get_resource)
    ) as mock:
        await tr.update()

        # Should catch if something happens with number of radar frames retrieved
        assert mock.call_count == 3

    assert test_radar == snapshot
