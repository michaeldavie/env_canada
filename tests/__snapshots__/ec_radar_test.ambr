# serializer version: 1
# name: test_get_radar_image_with_mock_data
  'test_radar()'
# ---
